### ISSUE TITLE
It's impossible to change the working directory from Rust

### ISSUE THREAD

**Issue by matthias-t** [matthias-t/workspace#10]
See [this question](https://stackoverflow.com/questions/43361154/how-to-change-os-to-directory-using-rust) on stackoverflow. It's also impossible to run commands that directly affect the user's current open shell. Using `std::process::Command` will spawn a new process, and changing the directory will only affect that new process.

### Now what?

The idea is to create shell functions that wrap workspace. They read the output, pick out lines that start with, say, `!`, and execute those, then print the rest. We could take inspiration from [gsamokovarov/jump](https://github.com/gsamokovarov/jump), they seem to follow a similar approach (see their shell folder).

**Comment by matthias-t** [comment]
Hours of trial and error on my Ubuntu WSL led to this:
```bash
function ws {
    while read -r; do 
        if [[ "$REPLY" == \!* ]]; then 
            source <(echo "${REPLY:1}"); 
        else
            echo $REPLY; 
        fi; 
    done < <( workspace );
}
```
@roma0615 does this work on your machine?

**Comment by roma0615** [comment]
Whats wrong with creating a new process and running the commands there? 

Also, where should I try to run that snippet from? Should it be from the same directory as the workspace executable?

**Comment by matthias-t** [comment]
The problem is that we need to change the working directory of the shell when a workspace is opened, and therefore we need to run commands in the shell, not a new process. If we run `cd /whatever` in a new process then it won't affect the shell.

If you wish, you can `cargo install --path .` (assuming you are in the directory where our `Cargo.toml` is) to install workspace on you machine. In my WSL I just created a 
```bash
function workspace { cat ~/test.txt; }
```
And I put some test lines in `~/test.txt`:
```
Hello world
!cd /
!ls
```

**Comment by matthias-t** [comment]
Here's the script for `cmd` (I would have honestly preferred writing it in Brainfuck if that was possible)
```cmd
@ECHO off
SETLOCAL enabledelayedexpansion
FOR /F "tokens=*" %%s IN ('workspace %*') DO (
    SET line=%%s
    IF "!line:~0,4!" == "RUN>" (
        CALL !line:~4!
        PUSHD .
        ENDLOCAL
        POPD
        SETLOCAL enabledelayedexpansion
    ) ELSE (
        ECHO !line!
    )
)
```
I only managed to make the path change affect the shell (with `pushd`/`popd` – it's kind of a hack, but hey, at least it works), other environment changes won't affect the shell process. I changed the prefix from `!` to `RUN>` because apparently `cmd` doesn't like `!` (*i.e.* won't read it from a file). It also fixes the colors and I have no idea why.
```
workspace list
```
![image](https://user-images.githubusercontent.com/26206173/41937611-11f3a358-7991-11e8-8433-3c933d02a8ae.png)
(I use cmder normally so I didn't notice)
```
testscript list
```
![image](https://user-images.githubusercontent.com/26206173/41937661-395f7df4-7991-11e8-909f-8a81187a472b.png)
